### PR TITLE
Enable use of a prompt engine within a closed-loop program/shell

### PR DESCRIPTION
**Pull request by maxlandon**
This little PR aims, as the title implies, to enable the use of oh-my-posh prompt engine within a closed-loop shell. All changes are described in the unique commit of this PR. Thanks for considering it

### CODE REVIEW

**Comment by maxlandon** [comment]
Note that the various changes to go.mod/go.sum are due to some automatic feature in my editor, and not caused by me adding some package in the PR: the changes in this file can thus probably be ignored safely.

**Comment by JanDeDobbeleer** [comment]
@maxlandon let me get this right, you want to use oh-my-posh as a library, right? If so, we should maybe also use `NewEngine` in our own code.

**Comment by maxlandon** [comment]
Hello !

Fastest answer ever !
Yes that is right, this is what I want.
To get some context: https://github.com/reeflective/readline
The way I use the prompt functions: https://github.com/reeflective/readline/blob/master/prompt.go

Available any time if you have questions or need help !
Thanks a lot !

**Comment by JanDeDobbeleer** [comment]
> Note that the various changes to go.mod/go.sum are due to some automatic feature in my editor

@maxlandon I'm going to revert those though, these shouldn't have been committed.

**Comment by maxlandon** [comment]
Yes sorry, I squashed my changes yesterday and completely forgot to do it !

File: src/engine/engine.go [hunk 1]
```diff
@@ -321,7 +332,7 @@ func (e *Engine) PrintTooltip(tip string) string {
 		Segments:  []*Segment{tooltip},
 	}
 	switch e.Env.Shell() {
-	case shell.ZSH, shell.CMD, shell.FISH:
+	case shell.ZSH, shell.CMD, shell.FISH, shell.PLAIN:
```
Comment (JanDeDobbeleer) [inline src/engine/engine.go:335]: why was this added? It's not really used anywhere as it's not a valid shell name.

File: src/engine/config.go [hunk 2]
```diff
@@ -218,7 +219,7 @@ func (cfg *Config) Write(format string) {
 	if len(destination) == 0 {
 		destination = cfg.origin
 	}
-	f, err := os.OpenFile(destination, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0755)
+	f, err := os.OpenFile(destination, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0o755)
```
Comment (JanDeDobbeleer) [inline src/engine/config.go:222]: this was a mistake, right?

File: src/engine/engine.go [hunk 3]
```diff
@@ -410,8 +421,10 @@ func (e *Engine) PrintExtraPrompt(promptType ExtraPromptType) string {
 			return prompt
 		}
 		return str
-	case shell.PWSH, shell.PWSH5, shell.CMD, shell.BASH, shell.FISH, shell.NU:
+	case shell.PWSH, shell.PWSH5, shell.CMD, shell.BASH, shell.FISH, shell.NU, shell.PLAIN:
```
Comment (JanDeDobbeleer) [inline src/engine/engine.go:424]: Same question as above

File: src/engine/new.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+)
+
+// NewEngine returns a prompt engine initialized with the
+// given configuration options, and is ready to print any
+// of the prompt components.
+func NewEngine(config string) *Engine {
+	env := &platform.Shell{
+		CmdFlags: &platform.Flags{
+			Config: config,
+			Shell:  "shell",
```
Comment (JanDeDobbeleer) [inline src/engine/new.go:17]: this needs to be exposed imho

**Comment by JanDeDobbeleer** [comment]
@maxlandon you can find it here #3223, I continued your work, but now we also use that API internally. If that works, I'll merge it and credit you.

File: src/engine/config.go (hunk 2 above)
Comment (maxlandon) [reply src/engine/config.go:222]: This is due to my editor applying to some linters.

File: src/engine/engine.go (hunk 1 above)
Comment (maxlandon) [reply src/engine/engine.go:335]: So this because if the `Shell()` call returns an empty string, the case statement is skipped and thus an empty prompt string is returned as well.

File: src/engine/engine.go (hunk 3 above)
Comment (maxlandon) [reply src/engine/engine.go:424]: Same response as above :)
However if there is something I missed in my approach, which results in this being useless, of course feel free to ignore it !

File: src/engine/new.go (hunk 4 above)
Comment (maxlandon) [reply src/engine/new.go:17]: Usually package writers expose functionality through some `pkg` package directory.
I have no particular suggestion as to where this should be exposed, you probably are better placed to decide.

File: src/engine/engine.go (hunk 1 above)
Comment (JanDeDobbeleer) [reply src/engine/engine.go:335]: That could make sense, I'll add a sane default in that case.
